# _Tasca 5, Sprint 3:_ *Dataframe Structures* 

## Exercise 1

Download the "Airlines Delay: Airline on-time statistics and delay causes" data set and upload it to a Dataframe pandas. Explore the data it contains, and keep only the columns you consider relevant.

In [1]:
# importing libraries
import pandas as pd
import numpy as np
pd.set_option ('display.max_columns', None)

In [2]:
pd.options.display.float_format = '{:.0f}'.format
filename = '../Data/DelayedFlights.csv'

data = pd.read_csv(filename, delimiter=',')

print('Shape of the data is', data.shape)
data.describe()

Shape of the data is (1936758, 30)


Unnamed: 0    Year   Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
count     1936758 1936758 1936758     1936758    1936758  1936758     1936758   
mean      3341651    2008       6          16          4     1519        1467   
std       2066065       0       3           9          2      450         425   
min             0    2008       1           1          1        1           0   
25%       1517452    2008       3           8          2     1203        1135   
50%       3242558    2008       6          16          4     1545        1510   
75%       4972467    2008       9          23          6     1900        1815   
max       7009727    2008      12          31          7     2400        2359   

       ArrTime  CRSArrTime  FlightNum  ActualElapsedTime  CRSElapsedTime  \
count  1929648     1936758    1936758            1928371         1936560   
mean      1610        1634       2184                133             134   
std        548         465       1945                 72              71   
min          1           0          1                 14             -25   
25%       1316        1325        610                 80              82   
50%       1715        1705       1543                116             116   
75%       2030        2014       3422                165             165   
max       2400        2400       9742               1114             660   

       AirTime  ArrDelay  DepDelay  Distance  TaxiIn  TaxiOut  Cancelled  \
count  1928371   1928371   1936758   1936758 1929648  1936303    1936758   
mean       108        42        43       766       7       18          0   
std         69        57        53       574       5       14          0   
min          0      -109         6        11       0        0          0   
25%         58         9        12       338       4       10          0   
50%         90        24        24       606       6       14          0   
75%        137        56        53       998       8       21          0   
max       1091      2461      2467      4962     240      422          1   

       Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
count   1936758       1247488       1247488   1247488        1247488   
mean          0            19             4        15              0   
std           0            44            21        34              2   
min           0             0             0         0              0   
25%           0             0             0         0              0   
50%           0             2             0         2              0   
75%           0            21             0        15              0   
max           1          2436          1352      1357            392   

       LateAircraftDelay  
count            1247488  
mean                  25  
std                   42  
min                    0  
25%                    0  
50%                    8  
75%                   33  
max                 1316

**_Data Exploration_**

The data has 1,936,758 rows and 30 columns. This is a data for Airlines delay statistics from the year 2008, and includes columns such as *departure time*, *arrival time*, *flight distance*, if the flights were *diverted and/or cancelled* and so on. Some of the columns may not be useful for the analysis concerning flight delays. Let's first see what are all the columns and the datatype for each of them. The built-in function of pandas called "*data.info*" will give us this information. 

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

Now, let's add a column called *'DepDate'* with Year-Month-Day format. This way I can remove the columns of 'Year' and 'DayofMonth'.

In [4]:
# 'Departure_date' with Year-Month-Day format
data['DepDate'] = pd.to_datetime(data.Year*10000+data.Month*100+data.DayofMonth,format='%Y%m%d')

**_Removal of columns_**

Now, Considering that the idea is to have an analysis for flight delays, the following columns can be deleted.  

    Unnamed: 0 - empty, just an index column
    Year - converted to date
    DayofMonth - converted to date
    DepTime, CRSDepTime - DepDelay is more useful
    ArrTime, CRSArrTime - ArrDelay is more useful
    FlightNum - carrier informations is more
    TailNum - a unique value, not useful, besides carrier information is more useful  
    ActualElapsedTime - it is a flight duration in minutes, does not impact flight delays. I will use flight distance instead of this to see the impact on flight delays, which is more robust.
    CRSElapsedTime - scheduled flight duration, again distance is a more robust parameter
    TaxiIn, TaxiOut - Taxi time before departure and after arrival, do not appear to impact significantly to flight delays in general. Also, 'AirTime' seems a slightly more useful factor to understand flight delays. 

In [5]:
drop_columns = ['Unnamed: 0', 'Year', 'DayofMonth', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 
                'FlightNum', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'TaxiIn', 'TaxiOut']
#data_new = data
data.drop(drop_columns, axis = 1, inplace = True)

In [6]:
print('Shape of the reduced data is', data.shape)
data.head(4)

Shape of the reduced data is (1936758, 18)


Month  DayOfWeek UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
0      1          4            WN      116       -14         8    IAD  TPA   
1      1          4            WN      113         2        19    IAD  TPA   
2      1          4            WN       76        14         8    IND  BWI   
3      1          4            WN       77        34        34    IND  BWI   

   Distance  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       810          0                N         0           nan           nan   
1       810          0                N         0           nan           nan   
2       515          0                N         0           nan           nan   
3       515          0                N         0             2             0   

   NASDelay  SecurityDelay  LateAircraftDelay    DepDate  
0       nan            nan                nan 2008-01-03  
1       nan            nan                nan 2008-01-03  
2       nan            nan                nan 2008-01-03  
3         0              0                 32 2008-01-03

After removing the columns that appear to be of no use, and adding a column for departure date, we have 18 columns in the dataset.

## Exercise 2

Make a complete report of the date set : 
     
     Summarize the columns of interest statistically
     Find how many missing data are per column
     Create new columns (average flight speed, whether late or not ...)
     Table of airlines with the most accumulated arrears
     What are the longest flights? And the most backward? Etc.

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 18 columns):
 #   Column             Dtype         
---  ------             -----         
 0   Month              int64         
 1   DayOfWeek          int64         
 2   UniqueCarrier      object        
 3   AirTime            float64       
 4   ArrDelay           float64       
 5   DepDelay           float64       
 6   Origin             object        
 7   Dest               object        
 8   Distance           int64         
 9   Cancelled          int64         
 10  CancellationCode   object        
 11  Diverted           int64         
 12  CarrierDelay       float64       
 13  WeatherDelay       float64       
 14  NASDelay           float64       
 15  SecurityDelay      float64       
 16  LateAircraftDelay  float64       
 17  DepDate            datetime64[ns]
dtypes: datetime64[ns](1), float64(8), int64(5), object(4)
memory usage: 266.0+ MB


**_Statistical summary of columns_**

I chose following columns where statistical summary may give useful information. 
    
    - AirTime, ArrDelay, DepDelay and Distance
    - CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay

In [18]:
col_of_interest = ['AirTime', 'ArrDelay', 'DepDelay', 'Distance', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']
data[col_of_interest].describe()

AirTime  ArrDelay  DepDelay  Distance  CarrierDelay  WeatherDelay  \
count  1928371   1928371   1936758   1936758       1247488       1247488   
mean       108        42        43       766            19             4   
std         69        57        53       574            44            21   
min          0      -109         6        11             0             0   
25%         58         9        12       338             0             0   
50%         90        24        24       606             2             0   
75%        137        56        53       998            21             0   
max       1091      2461      2467      4962          2436          1352   

       NASDelay  SecurityDelay  LateAircraftDelay  
count   1247488        1247488            1247488  
mean         15              0                 25  
std          34              2                 42  
min           0              0                  0  
25%           0              0                  0  
50%           2              0                  8  
75%          15              0                 33  
max        1357            392               1316

The following observations can be made from this summary:

    - The average time for flights in air is 108 min and the maximum flight time is 1091 minutes. 
    - The average distance travelled is 766 miles, whereas the maximum distance is 4962 miles. 
    - The average arrival and departure delays are 42 and 43 minutes, respectively. 
    - The minimum times for departure delay is 6 minutes and for arrival delay is -109 minutes, the later is puzzling because it indicates that the flight arrived 109 minutes before scheduled arrival time. There are also other negative arrival delays in the column. 
    - From the remaining columns, we can see that the carrier delays (due to airline) is a major reason for delays (2436), while security related delays lead to least number of flight delays (392). 

**_Missing data per columns_**

To find out how many missing values are present in the data, the built-in function of pandas called "*isnull*" is used. Adding the number of times a missing data is detected over all the rows and columns, we can get total number of missing values. To get the missing values in each column, the detected values are summed once (only over columns).

In [9]:
nan_info=pd.DataFrame(data.dtypes).T.rename(index={0:'column type'})

nan_info=nan_info.append(pd.DataFrame(data.isnull().sum()).T.rename(index={0:'null values (nb)'}))

nan_info=nan_info.append(pd.DataFrame(data.isnull().sum()/data.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
nan_info.T

column type null values (nb) null values (%)
Month                       int64                0               0
DayOfWeek                   int64                0               0
UniqueCarrier              object                0               0
AirTime                   float64             8387               0
ArrDelay                  float64             8387               0
DepDelay                  float64                0               0
Origin                     object                0               0
Dest                       object                0               0
Distance                    int64                0               0
Cancelled                   int64                0               0
CancellationCode           object                0               0
Diverted                    int64                0               0
CarrierDelay              float64           689270              36
WeatherDelay              float64           689270              36
NASDelay                  float64           689270              36
SecurityDelay             float64           689270              36
LateAircraftDelay         float64           689270              36
DepDate            datetime64[ns]                0               0

**_Adding New columns_**

Now, let's add some new columns to the data to give some extra information regarding flight delay statistics.

    "AvgSpeed": Average flight speed, calculated with distance divided by airtime (converted to miles/hour)
    "Delay": Delay status: on time (0), delayed (1), if flight is more than 15 minutes late, then it is considered   late, otherwise on time
    "FLstatus": Flight status, on time (0), slightly delayed (1), highly delayed (2), diverted (3), cancelled (4)
    

In [10]:
data['AvgSpeed'] = data['Distance'] / data['AirTime']*60 # the unit is miles/hour
#data.head(1)

In [11]:
for dataset in data:
    data.loc[data['ArrDelay'] <= 15, 'Delay'] = 0
    data.loc[data['ArrDelay'] >= 15, 'Delay'] = 1

In [12]:
for dataset in data:
    data.loc[data['ArrDelay'] <= 15, 'FLstatus'] = 0
    data.loc[data['ArrDelay'] >= 15, 'FLstatus'] = 1
    data.loc[data['ArrDelay'] >= 60, 'FLstatus'] = 2
    data.loc[data['Diverted'] == 1, 'FLstatus'] = 3
    data.loc[data['Cancelled'] == 1, 'FLstatus'] = 4

In [13]:
print('Shape of the data with added columns is', data.shape)
data.head(4)

Shape of the data with added columns is (1936758, 21)


Month  DayOfWeek UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
0      1          4            WN      116       -14         8    IAD  TPA   
1      1          4            WN      113         2        19    IAD  TPA   
2      1          4            WN       76        14         8    IND  BWI   
3      1          4            WN       77        34        34    IND  BWI   

   Distance  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       810          0                N         0           nan           nan   
1       810          0                N         0           nan           nan   
2       515          0                N         0           nan           nan   
3       515          0                N         0             2             0   

   NASDelay  SecurityDelay  LateAircraftDelay    DepDate  AvgSpeed  Delay  \
0       nan            nan                nan 2008-01-03       419      0   
1       nan            nan                nan 2008-01-03       430      0   
2       nan            nan                nan 2008-01-03       407      0   
3         0              0                 32 2008-01-03       401      1   

   FLstatus  
0         0  
1         0  
2         0  
3         1

Now that we have converted *Diverted* and *Cancelled* to Flight_status, we can delete these two columns. 

In [14]:
data = data.drop("Cancelled",1)
data = data.drop("Diverted",1)

In [15]:
print('Shape of the data is', data.shape)
data.head(4)

Shape of the data is (1936758, 19)


Month  DayOfWeek UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
0      1          4            WN      116       -14         8    IAD  TPA   
1      1          4            WN      113         2        19    IAD  TPA   
2      1          4            WN       76        14         8    IND  BWI   
3      1          4            WN       77        34        34    IND  BWI   

   Distance CancellationCode  CarrierDelay  WeatherDelay  NASDelay  \
0       810                N           nan           nan       nan   
1       810                N           nan           nan       nan   
2       515                N           nan           nan       nan   
3       515                N             2             0         0   

   SecurityDelay  LateAircraftDelay    DepDate  AvgSpeed  Delay  FLstatus  
0            nan                nan 2008-01-03       419      0         0  
1            nan                nan 2008-01-03       430      0         0  
2            nan                nan 2008-01-03       407      0         0  
3              0                 32 2008-01-03       401      1         1

In [16]:
data.rename(columns={'UniqueCarrier': 'Airline'}, inplace=True)
data.rename(columns={'Dest': 'Destination'}, inplace=True)
data.head(2)

Month  DayOfWeek Airline  AirTime  ArrDelay  DepDelay Origin Destination  \
0      1          4      WN      116       -14         8    IAD         TPA   
1      1          4      WN      113         2        19    IAD         TPA   

   Distance CancellationCode  CarrierDelay  WeatherDelay  NASDelay  \
0       810                N           nan           nan       nan   
1       810                N           nan           nan       nan   

   SecurityDelay  LateAircraftDelay    DepDate  AvgSpeed  Delay  FLstatus  
0            nan                nan 2008-01-03       419      0         0  
1            nan                nan 2008-01-03       430      0         0

## Exercise 3

Export the data set clean and with the new columns to Excel.

In [17]:
data.to_csv('../Data/FlightDelay_new.csv')  